common/lib/plugins/efm2/monitor.ts

```diff
@@ -0,0 +1,310 @@
+   *
+   * @return whether the server is still alive and the elapsed time spent checking.
+   */
+  async checkConnectionStatus(): Promise<boolean> {
+    const connectContext = this.telemetryFactory.openTelemetryContext("Connection status check", TelemetryTraceLevel.FORCE_TOP_LEVEL);
+    connectContext.setAttribute("url", this.hostInfo.host);
+    try {
+      if (!(await this.pluginService.isClientValid(this.monitoringClient))) {
+        // Open a new connection.
+        const monitoringConnProperties: Map<string, any> = new Map(this.properties);
+
+        for (const key of this.properties.keys()) {
+          if (!key.startsWith(WrapperProperties.MONITORING_PROPERTY_PREFIX)) {
+            continue;
+          }
+
+          monitoringConnProperties.set(key.substring(WrapperProperties.MONITORING_PROPERTY_PREFIX.length), this.properties.get(key));
+          monitoringConnProperties.delete(key);
+        }
+
+        logger.debug(`Opening a monitoring connection to ${this.hostInfo.url}`);
+        this.monitoringClient = await this.pluginService.forceConnect(this.hostInfo, monitoringConnProperties);
+        logger.debug(`Successfully opened monitoring connection to ${this.monitoringClient.id} - ${this.hostInfo.url}`);
+        return true;
+      }
+      return true;
+    } catch (error: any) {
+      return false;
+    }
+  }
+
+  isStopped(): boolean {
+    return this.stopped;
+  }
+
+  async end(): Promise<void> {
+    this.stopped = true;
+    // Waiting for 30s gives a task enough time to exit monitoring loop and close database connection.
+    await sleep(30000);
+    logger.debug(Messages.get("MonitorImpl.stopped", this.hostInfo.host));
+  }
+
+  updateHostHealthStatus(connectionValid: boolean, statusCheckStartNano: number, statusCheckEndNano: number): Promise<void> {
+    if (!connectionValid) {
+      this.failureCount += 1;
+
+      if (this.invalidHostStartTimeNano === 0) {
+        this.invalidHostStartTimeNano = statusCheckStartNano;
+      }
+
+      const invalidHostDurationNano = statusCheckEndNano - this.invalidHostStartTimeNano;
+      const maxInvalidHostDurationNano = this.failureDetectionIntervalNanos * Math.max(0, this.failureDetectionCount - 1);
+
+      if (invalidHostDurationNano >= maxInvalidHostDurationNano) {
+        logger.debug(Messages.get("MonitorConnectionContext.hostDead", this.hostInfo.host));
+        this.hostUnhealthy = true;
+        return Promise.resolve();
+      }
+      logger.debug(Messages.get("MonitorConnectionContext.hostNotResponding", this.hostInfo.host, this.failureCount.toString()));
+      return Promise.resolve();
+    }
+
+    if (this.failureCount > 0) {
+      // Host is back alive
+      logger.debug(Messages.get("MonitorConnectionContext.hostAlive", this.hostInfo.host));
+    }
+    this.failureCount = 0;
+    this.invalidHostStartTimeNano = 0;
+    this.hostUnhealthy = false;
+  }
+
+  protected getCurrentTimeNano() {
```

Comment:
This method has been defined before in `efm/monitor.ts`, maybe we can put this in `utils`